tests/unit/asio_service_stream_test.cxx

```diff
@@ -29,57 +29,178 @@ using namespace raft_functional_common;
 namespace asio_service_stream_test {
     const std::string TEST_MSG = "stream-test-msg-str";
 
+    class stream_statistic {
+    public:
+        stream_statistic() 
+        : resp_log_index_(0)
+        , msg_mismatch_(false)
+        , reqs_out_of_order_(false)
+        , next_log_index_(1)
+        {
+            std::string server_log_file_name = "./srv" + std::to_string(1) + ".log";
+            server_logger_ = cs_new<logger_wrapper>(server_log_file_name);
+            std::string client_log_file_name = "./srv" + std::to_string(2) + ".log";
+            client_logger_ = cs_new<logger_wrapper>(client_log_file_name);
+        }
+
+        bool waiting_for_responses(int timeout_ms = 3000) {
+            TestSuite::_msg("wait for responses (up to %d ms)\n", timeout_ms);
+            ea_.wait_ms(timeout_ms);
+            SimpleLogger* ll = client_logger_->getLogger();
+            _log_info(ll, "resp: %ld, message sent: %ld", next_log_index_.load(), num_messages_sent_);
+            return (next_log_index_ == num_messages_sent_ + 1);
+        }
+
+        void wait_for_receiving_requests(int timeout_ms = 1000) {
+            TestSuite::_msg("wait for receiving requests (up to %d ms)\n", timeout_ms);
+            req_ea_.wait_ms(1000);
+        }
+
+        // server
+        std::atomic<ulong> resp_log_index_;
+        std::atomic<bool> msg_mismatch_;
+        std::atomic<bool> reqs_out_of_order_;
+        EventAwaiter req_ea_;
+
+        // client
+        std::atomic<ulong> next_log_index_;
+        size_t sent_log_index_ = 0;
+        size_t num_messages_sent_ = 0;
+        size_t error_req_count_ = 0;
+        EventAwaiter ea_;
+
+        // log
+        ptr<logger_wrapper> server_logger_;
+        ptr<logger_wrapper> client_logger_;
+    };
+
     class stream_msg_handler : public nuraft::msg_handler {
     public:
         stream_msg_handler(context* ctx,
                            const init_options& opt,
-                           ptr<logger_wrapper> log_wrapper)
+                           ptr<stream_statistic> stream_stat)
```

Comment:
If `stream_statistic` is shared throughout the test, why don't you just define it as a global variable?
```c++
static stream_statistic stream_stat;
```
And reset it at the beginning of each test, by implementing `reset()` API.
```c++
stream_stat.reset();
```
Need to make sure `ea_.reset()` should be invoked before reusing it.